File: paper_trading_ui/backend/schemas/accounts.py
from __future__ import annotations

from pydantic import BaseModel


class AccountParamsRequest(BaseModel):
    strategy: str | None = None
    accountKind: str | None = None
    descriptiveName: str | None = None
    riskPolicy: str | None = None
    stopLossPct: float | None = None
    takeProfitPct: float | None = None
    tradeSizePct: float | None = None
    maxPositionPct: float | None = None
    instrumentMode: str | None = None
    goalMinReturnPct: float | None = None
    goalMaxReturnPct: float | None = None
    goalPeriod: str | None = None
    learningEnabled: bool | None = None
    optionStrikeOffsetPct: float | None = None
    optionMinDte: int | None = None
    optionMaxDte: int | None = None
    optionType: str | None = None
    targetDeltaMin: float | None = None
    targetDeltaMax: float | None = None
    maxPremiumPerTrade: float | None = None
    maxContractsPerTrade: int | None = None
    ivRankMin: float | None = None
    ivRankMax: float | None = None
    rollDteThreshold: int | None = None
    profitTakePct: float | None = None
    maxLossPct: float | None = None
    rotationEnabled: bool | None = None
    rotationMode: str | None = None
    rotationOptimalityMode: str | None = None
    rotationIntervalDays: int | None = None
    rotationIntervalMinutes: int | None = None
    rotationLookbackDays: int | None = None
    rotationSchedule: list[str] | None = None
    rotationRegimeStrategyRiskOn: str | None = None
    rotationRegimeStrategyNeutral: str | None = None
    rotationRegimeStrategyRiskOff: str | None = None
    rotationOverlayMode: str | None = None
    rotationOverlayMinTickers: int | None = None
    rotationOverlayConfidenceThreshold: float | None = None
    rotationOverlayWatchlist: list[str] | None = None
    rotationActiveIndex: int | None = None
    rotationLastAt: str | None = None
    rotationActiveStrategy: str | None = None

File: apps/paper_trading_ui/frontend/src/types/accounts.ts
import type { BacktestRunSummary, LatestBacktestMetrics } from "./backtesting";

export type AccountListItem = {
  name: string;
  displayName: string;
  strategy: string;
  instrumentMode: string;
  benchmark: string;
  equity: number;
  totalChange: number;
  totalChangePct: number;
  changeSinceLastSnapshot: number | null;
  latestSnapshotTime: string | null;
};

export interface AccountConfigFields {
  riskPolicy: string;
  stopLossPct: number | null;
  takeProfitPct: number | null;
  tradeSizePct: number;
  maxPositionPct: number;
  instrumentMode: string;
  goalMinReturnPct: number | null;
  goalMaxReturnPct: number | null;
  goalPeriod: string | null;
  learningEnabled: boolean;
  optionStrikeOffsetPct: number | null;
  optionMinDte: number | null;
  optionMaxDte: number | null;
  optionType: string | null;
  targetDeltaMin: number | null;
  targetDeltaMax: number | null;
  maxPremiumPerTrade: number | null;
  maxContractsPerTrade: number | null;
  ivRankMin: number | null;
  ivRankMax: number | null;
  rollDteThreshold: number | null;
  profitTakePct: number | null;
  maxLossPct: number | null;
  rotationEnabled?: boolean;
  rotationMode?: string;
  rotationOptimalityMode?: string;
  rotationIntervalDays?: number | null;
  rotationIntervalMinutes?: number | null;
  rotationLookbackDays?: number | null;
  rotationSchedule?: string[] | null;
  rotationRegimeStrategyRiskOn?: string | null;
  rotationRegimeStrategyNeutral?: string | null;
  rotationRegimeStrategyRiskOff?: string | null;
  rotationOverlayMode?: string;
  rotationOverlayMinTickers?: number | null;
  rotationOverlayConfidenceThreshold?: number | null;
  rotationOverlayWatchlist?: string[] | null;
  rotationActiveIndex?: number | null;
  rotationLastAt?: string | null;
  rotationActiveStrategy?: string | null;
}

export interface AccountMutableIdentityFields {
  strategy: string;
  descriptiveName: string;
}

export interface AccountConfigOptionDefaults {
  goalPeriod: string;
  riskPolicy: string;
  instrumentMode: string;
  rotationMode: string;
  rotationOptimalityMode: string;
  rotationOverlayMode: string;
}

export interface AccountConfigOptions {
  goalPeriods: string[];
  riskPolicies: string[];
  instrumentModes: string[];
  optionTypes: string[];
  rotationModes: string[];
  rotationOptimalityModes: string[];
  rotationOverlayModes: string[];
  defaults: AccountConfigOptionDefaults;
}

export type AccountSummary = AccountListItem & AccountConfigFields & {
  initialCash: number;
  settlementCash: number;
  liveBenchmarkReturnPct: number | null;
  liveAlphaPct: number | null;
  liveBenchmarkEquity: number | null;
  liveBenchmarkStartTime: string | null;
  liveBenchmarkEndTime: string | null;
};

export type LiveBenchmarkOverlay = {
  benchmark: string;
  startTime: string;
  endTime: string;
  startingEquity: number;
  endingEquity: number;
  benchmarkEquity: number;
  accountReturnPct: number;
  benchmarkReturnPct: number;
  alphaPct: number;
  points: Array<{
    time: string;
    accountEquity: number;
    benchmarkEquity: number;
  }>;
};

export type AccountDetail = {
  account: AccountSummary;
  latestBacktest: BacktestRunSummary | null;
  latestBacktestMetrics?: LatestBacktestMetrics | null;
  liveBenchmarkOverlay?: LiveBenchmarkOverlay | null;
  snapshots: Array<{
    time: string;
    cash: number;
    marketValue: number;
    equity: number;
    realizedPnl: number;
    unrealizedPnl: number;
  }>;
  trades: Array<{
    ticker: string;
    side: string;
    qty: number;
    price: number;
    fee: number;
    tradeTime: string;
    note: string | null;
  }>;
  positions: Array<{
    ticker: string;
    qty: number;
    avgCost: number;
    marketPrice: number;
    marketValue: number;
    unrealizedPnl: number;
  }>;
};

export type AnalysisPosition = {
  ticker: string;
  qty: number;
  avgCost: number;
  costBasis: number;
  marketPrice: number;
  marketValue: number;
  unrealizedPnl: number;
  unrealizedPnlPct: number;
  portfolioPct: number;
};

export type AccountAnalysis = {
  accountReturnPct: number;
  benchmarkReturnPct: number | null;
  benchmarkTicker: string | null;
  alphaPct: number | null;
  realizedPnl: number;
  unrealizedPnl: number;
  equity: number;
  topWinners: AnalysisPosition[];
  topLosers: AnalysisPosition[];
  improvementNotes: string[];
};

export type AccountParamsUpdate = Partial<AccountConfigFields> & Partial<AccountMutableIdentityFields>;
